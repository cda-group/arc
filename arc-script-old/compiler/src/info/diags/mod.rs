/// Module for representing diagnostics.
pub mod repr;
/// Module which exposes structs which diagnostics can be written to.
pub mod sink;
/// Module for transforming [`repr::Diagnostic`] into [`codespan_reporting::diagnostic::Diagnostic`].
pub mod to_codespan;

pub use repr::*;
pub use sink::*;
