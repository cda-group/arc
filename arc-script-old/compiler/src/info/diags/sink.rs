use arc_script_compiler_shared::New;

pub use codespan_reporting::term::termcolor::Buffer;
pub use codespan_reporting::term::termcolor::ColorChoice;
pub use codespan_reporting::term::termcolor::StandardStream;
pub use codespan_reporting::term::termcolor::WriteColor;

use std::io::Result;

use codespan_reporting::term::termcolor::ColorSpec;

/// Trait which wraps `Write` and `WriteColor`.
pub trait Writer: std::io::Write + WriteColor {}

impl<T> Writer for T where T: std::io::Write + WriteColor {}

/// A writer which will move data into the void.
/// Just like [`std::io::Sink`], but supports the `termcolor` library.
#[derive(Debug, Default, New)]
pub struct Sink;

impl WriteColor for Sink {
    fn supports_color(&self) -> bool {
        true
    }

    fn set_color(&mut self, _: &ColorSpec) -> Result<()> {
        Ok(())
    }

    fn reset(&mut self) -> Result<()> {
        Ok(())
    }
}

impl std::io::Write for Sink {
    fn write(&mut self, _: &[u8]) -> Result<usize> {
        Ok(0)
    }

    fn flush(&mut self) -> Result<()> {
        Ok(())
    }
}
