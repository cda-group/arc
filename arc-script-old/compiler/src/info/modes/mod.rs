//! Meta-module for modes. The compiler can be run in different modes.
//! A mode determines how the compiler operates.

/// Module for representing modes.
pub mod repr;

pub use repr::*;
