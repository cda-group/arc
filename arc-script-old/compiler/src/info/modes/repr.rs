use arc_script_compiler_shared::Educe;

/// Compiler settings.
#[derive(Debug, Clone, Default)]
pub struct Mode {
    /// Activates debug mode when true. Debug mode will cause the compiler to emit internal messages
    /// that are useful for pinning down and squashing bugs within the compiler.
    pub debug: bool,
    /// If true, activate profiling mode. Profiling mode measures and emits the execution time of each
    /// compilation phase.
    pub profile: bool,
    /// Configures the verbosity level of the compiler. The levels correspond to [`tracing::Level`].
    pub verbosity: Verbosity,
    /// Suppresses diagnostics from being printed. Can be useful when using the compiler as a library.
    pub suppress_diags: bool,
    /// Fails compilation as soon as a compilation phase emits an error.
    pub fail_fast: bool,
    /// Skip type inference pass.
    pub no_infer: bool,
    /// Do not include prelude.
    pub no_prelude: bool,
    /// Forces the compiler to emit its internal representation (e.g., AST) even if it has errors.
    pub force_output: bool,
    /// Configures what kind of language the compiler expects its input source code to be written in.
    pub lang: Lang,
    /// Configures what kind of input the compiler reads.
    pub input: Input,
    /// Configures what kind of output the compiler writes.
    pub output: Output,
}

#[derive(Debug, Clone)]
/// Language modes.
pub enum Lang {
    /// Arc-Query
    Arq,
    /// Arc-Script
    Arc,
}

impl Default for Lang {
    fn default() -> Self {
        Self::Arc
    }
}

/// Verbosity levels
#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
pub enum Verbosity {
    /// Designates very serious errors.
    Error,
    /// Designates hazardous situations.
    Warn,
    /// Designates useful information.
    Info,
    /// Designates lower priority information.
    Debug,
    /// Designates very low priority information.
    Trace,
}

impl Default for Verbosity {
    fn default() -> Self {
        Self::Error
    }
}

/// The compiler can read input from different sources.
#[derive(Debug, Clone, Educe)]
#[educe(Default)]
pub enum Input {
    /// Read code directly from a source string.
    Code(String),
    /// Read input from a source file or root directory.
    #[cfg(not(target_arch = "wasm32"))]
    File(Option<std::path::PathBuf>),
    /// Input will be added later.
    #[educe(Default)]
    Empty,
}

/// The compiler can write
#[derive(Debug, Clone, Copy, Educe)]
#[educe(Default)]
pub enum Output {
    /// Emit [`crate::ast::AST`] as output (by first parsing it from source code).
    AST,
    /// Emit [`crate::hir::HIR`] as output (by first lowering the `AST` into it).
    HIR,
    /// Emit [`crate::rust::Rust`] as output (by first lowering the `HIR` into it).
    Rust,
    /// Emit Rust as output via the MLIR route.
    RustMLIR,
    /// Emit [`crate::mlir::MLIR`] as output (by first lowering the `HIR` into it).
    #[educe(Default)]
    MLIR,
    /// Emit no output.
    Silent,
}

/// Consult environment variables to determine the backend to use for
/// rust output.
#[must_use]
pub fn get_rust_backend() -> Output {
    let use_mlir_backend = match std::env::var("ARCSCRIPT_MLIR_BACKEND") {
        Ok(val) => val.parse().unwrap_or(0) != 0,
        Err(e) => false,
    };

    if use_mlir_backend {
        Output::RustMLIR
    } else {
        Output::Rust
    }
}
