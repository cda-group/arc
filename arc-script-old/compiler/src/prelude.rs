//! Exports for users who use the compiler as a library.

pub use crate::ast;
pub use crate::info::diags;
pub use crate::info::modes;
pub use arc_script_compiler_shared::Result;
