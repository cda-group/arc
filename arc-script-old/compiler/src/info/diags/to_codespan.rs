use crate::hir;
use crate::hir::HIR;
use crate::info::diags::Diagnostic;
use crate::info::diags::Error;
use crate::info::diags::Note;
use crate::info::diags::Panic;
use crate::info::diags::Warning;
use crate::info::files::FileId;
use crate::info::files::Loc;
use crate::info::Info;

use arc_script_compiler_shared::Shrinkwrap;

use codespan_reporting::diagnostic;
use codespan_reporting::diagnostic::Label;

use std::borrow::Borrow;
use std::str;

/// A [`codespan`] diagnostic.
pub type Codespan = diagnostic::Diagnostic<FileId>;

/// A struct which contains everything necessary to print diagnostics.
/// It is intended to be used by users of the `arc_script` compiler-library.
#[derive(Debug)]
pub struct Report {
    info: Info,
    hir: Option<HIR>,
}

impl Report {
    /// Constructs a new `Report` by taking ownership of the Arc-Script compiler's output.
    /// Reports can only be constructed within the compiler, therefore this is `pub(crate)`.
    pub(crate) const fn semantic(info: Info, hir: HIR) -> Self {
        Self {
            info,
            hir: Some(hir),
        }
    }
    pub(crate) const fn syntactic(info: Info) -> Self {
        Self { info, hir: None }
    }
    /// Returns `true` if there are no diagnostics to be reported, otherwise `false`.
    pub fn is_ok(&self) -> bool {
        self.info.diags.is_empty()
    }
}

impl From<Report> for (Vec<Codespan>, Info) {
    fn from(report: Report) -> Self {
        let Report { mut info, hir } = report;
        let diags = info.diags.take();
        let ctx = &Context {
            hir: hir.as_ref(),
            info: &info,
        };
        let codepan = diags
            .store
            .into_iter()
            .filter_map(|diag| diag.to_codespan(ctx))
            .collect();
        (codepan, info)
    }
}

/// Hacky solution, HIR can optionally be omitted to
/// print diagnostics from before it was constructed.
#[derive(Debug, Copy, Clone, Shrinkwrap)]
pub struct Context<'i> {
    #[shrinkwrap(main_field)]
    pub(crate) info: &'i Info,
    pub(crate) hir: Option<&'i HIR>,
}

impl<'i> Context<'i> {
    /// Constructs a context which can be used for converting into diagnostics.
    pub(crate) const fn new(info: &'i Info, hir: Option<&'i HIR>) -> Self {
        Self { info, hir }
    }
}

/// Integration with [`codespan_reporting`].
pub trait ToCodespan {
    /// Converts `Self` into a pretty `Codespan` diagnostic.
    fn to_codespan(&self, ctx: &Context<'_>) -> Option<Codespan>;
}

impl ToCodespan for Diagnostic {
    fn to_codespan(&self, ctx: &Context<'_>) -> Option<Codespan> {
        match self {
            Diagnostic::Error(diag) => diag.to_codespan(ctx),
            Diagnostic::Warning(diag) => diag.to_codespan(ctx),
            Diagnostic::Note(diag) => diag.to_codespan(ctx),
        }
    }
}

impl ToCodespan for Note {
    fn to_codespan(&self, _ctx: &Context<'_>) -> Option<Codespan> {
        crate::todo!()
    }
}

impl ToCodespan for Warning {
    fn to_codespan(&self, _ctx: &Context<'_>) -> Option<Codespan> {
        crate::todo!()
    }
}

impl ToCodespan for Error {
    fn to_codespan(&self, ctx: &Context<'_>) -> Option<Codespan> {
        match self {
            Error::FileNotFound => Codespan::error().with_message("Source file not found."),
            Error::BadExtension => {
                Codespan::error().with_message("Input files must have `.arc` extension.")
            }
            Error::ExtraToken { found, loc } => Codespan::error()
                .with_message(format!("Extraneous token {}", found.pretty(ctx.info)))
                .with_labels(vec![label(loc)?]),
            Error::InvalidToken { loc } => Codespan::error()
                .with_message("Invalid token")
                .with_labels(vec![label(loc)?]),
            Error::UnrecognizedEOF { loc, expected } => Codespan::error()
                .with_message("Unrecognized end of file")
                .with_labels(vec![
                    label(loc)?.with_message(format!("expected {}", expected.join(", ")))
                ]),
            Error::UnrecognizedToken {
                found,
                expected,
                loc,
            } => Codespan::error()
                .with_message(format!("Unrecognized token {}", found.pretty(ctx.info)))
                .with_labels(vec![
                    label(loc)?.with_message(format!("expected {}", expected.join(", ")))
                ]),
            Error::TypeMismatch { lhs, rhs, loc } => Codespan::error()
                .with_message("Type mismatch")
                .with_labels(vec![label(loc)?.with_message(format!(
                    "{} != {}",
                    ctx.hir.unwrap().pretty(lhs, ctx.info),
                    ctx.hir.unwrap().pretty(rhs, ctx.info)
                ))]),
            Error::PathNotFound { path, loc } => Codespan::error()
                .with_message(format!(
                    "Identifier `{}` not bound to anything",
                    ctx.hir.unwrap().pretty(path, ctx.info),
                ))
                .with_labels(vec![label(loc)?.with_message("Used here")]),
            Error::NonExhaustiveMatch { loc } => Codespan::error()
                .with_message("Match is non-exhaustive")
                .with_labels(vec![label(loc)?.with_message("Missing cases")]),
            Error::NameClash { name } => {
                Codespan::error()
                    .with_message("Name clash")
                    .with_labels(vec![label(name.loc)?
                        .with_message("Name shadows previous declaration".to_owned())])
            }
            Error::FieldClash { name } => Codespan::error()
                .with_message("Found duplicate key")
                .with_labels(vec![label(name.loc)?.with_message(format!(
                    "{}",
                    ctx.hir.unwrap().pretty(&name.id, ctx.info)
                ))]),
            Error::VariantClash { name } => Codespan::error()
                .with_message("Found duplicate key")
                .with_labels(vec![label(name.loc)?.with_message(format!(
                    "{}",
                    ctx.hir.unwrap().pretty(&name.id, ctx.info)
                ))]),
            Error::VariantWrongArity { path } => Codespan::error()
                .with_message("Variant constructors expect exactly one argument.")
                .with_labels(vec![label(path.loc)?.with_message(format!(
                    "{}",
                    ctx.hir.unwrap().pretty(path, ctx.info)
                ))]),
            Error::OutOfBoundsProject { loc } => Codespan::error()
                .with_message("Out of bounds projection")
                .with_labels(vec![label(loc)?]),
            Error::FieldNotFound { loc } => Codespan::error()
                .with_message("Field not found")
                .with_labels(vec![label(loc)?]),
            Error::TypeInValuePosition { loc } => Codespan::error()
                .with_message("Expected value, found type in value-position")
                .with_labels(vec![label(loc)?]),
            Error::ValueInTypePosition { loc } => Codespan::error()
                .with_message("Expected value, found type in value-position")
                .with_labels(vec![label(loc)?]),
            Error::RefutablePattern { loc } => Codespan::error()
                .with_message("Invalid pattern")
                .with_labels(vec![label(loc)?]),
            Error::TooMuchIndent { loc } => Codespan::error()
                .with_message("Too many levels of indentation. Consider un-indenting.")
                .with_labels(vec![label(loc)?]),
            Error::BadIndent { loc } => Codespan::error()
                .with_message("Incorrectly aligned indentation.")
                .with_labels(vec![label(loc)?]),
            Error::LexicalCore { err, loc } => Codespan::error()
                .with_message(lex_err(err))
                .with_labels(vec![label(loc)?]),
            Error::Time { err, loc } => Codespan::error()
                .with_message(err.to_string())
                .with_labels(vec![label(loc)?]),
            Error::UseOfMovedValue { loc0, loc1, t } => Codespan::error()
                .with_message(format!(
                    "Use of moved value, where moved value is of non-copyable type {}",
                    ctx.hir.unwrap().pretty(t, ctx.info)
                ))
                .with_labels(vec![label(loc0)?, label(loc1)?]),
            Error::DoubleUse {
                loc0,
                loc1,
                loc2,
                t,
            } => Codespan::error()
                .with_message(format!(
                    "Double use of value of non-copyable type {}",
                    ctx.hir.unwrap().pretty(t, ctx.info)
                ))
                .with_labels(vec![label(loc0)?, label(loc1)?, label(loc2)?]),
            Error::PathIsNotVariant { loc } => Codespan::error()
                .with_message("Path is not referring to a variant")
                .with_labels(vec![label(loc)?]),
            Error::PatternInExternFun { loc } => Codespan::error()
                .with_message("Extern functions may only take parameters which are not patterns")
                .with_labels(vec![label(loc)?]),
            Error::ExpectedSelector { loc } => Codespan::error()
                .with_message("Expected selector")
                .with_labels(vec![label(loc)?]),
            Error::MultipleSelectors { loc } => Codespan::error()
                .with_message("Found multiple selectors, expected just one.")
                .with_labels(vec![label(loc)?]),
            Error::ExpectedSelectableType { loc } => Codespan::error()
                .with_message("Expected selectable type.")
                .with_labels(vec![label(loc)?]),
            Error::TypeMustBeKnownAtThisPoint { loc } => Codespan::error()
                .with_message("Type must be known at this point.")
                .with_labels(vec![label(loc)?]),
        }
        .into()
    }
}

/// Returns a textual representation of an error produced by [`lexical_core`].
#[rustfmt::skip]
fn lex_err(err: &lexical_core::Error) -> &'static str {
    use lexical_core::ErrorCode::*;
    match err.code {
        Overflow                    => "Integral overflow occurred during numeric parsing.",
        Underflow                   => "Integral underflow occurred during numeric parsing.",
        InvalidDigit                => "Invalid digit found before string termination.",
        Empty                       => "Empty byte array found.",
        EmptyMantissa               => "Empty mantissa found.",
        EmptyExponent               => "Empty exponent found.",
        EmptyInteger                => "Empty integer found.",
        EmptyFraction               => "Empty fraction found.",
        InvalidPositiveMantissaSign => "Invalid positive mantissa sign was found.",
        MissingMantissaSign         => "Mantissa sign was required, but not found.",
        InvalidExponent             => "Exponent was present but not allowed.",
        InvalidPositiveExponentSign => "Invalid positive exponent sign was found.",
        MissingExponentSign         => "Exponent sign was required, but not found.",
        ExponentWithoutFraction     => "Exponent was present without fraction component.",
        InvalidLeadingZeros         => "Integer had invalid leading zeros.",
        __Nonexhaustive             => crate::todo!(),
    }
}

fn label(loc: impl Borrow<Loc>) -> Option<Label<FileId>> {
    if let Loc::Real(file, span) = loc.borrow() {
        Some(Label::primary(*file, *span))
    } else {
        None
    }
}
