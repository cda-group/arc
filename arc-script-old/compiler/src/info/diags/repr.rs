use crate::ast::lower::source::lexer::sem_tokens::Token;
use crate::ast::Name;

use crate::hir::Path;
use crate::hir::HIR;
use crate::info::diags::to_codespan::Context;
use crate::info::diags::to_codespan::ToCodespan;
use crate::info::files::Loc;

use crate::info::types::Type;
use crate::info::Info;

use std::io::Write;

use codespan_reporting::term;

use codespan_reporting::term::termcolor::Color;

use codespan_reporting::term::termcolor::ColorSpec;

use arc_script_compiler_shared::From;
use arc_script_compiler_shared::Shrinkwrap;
use codespan_reporting::term::termcolor::WriteColor;
use codespan_reporting::term::Config;

type CodespanResult = std::result::Result<(), codespan_reporting::files::Error>;

/// Interner for storing diagnostics.
#[derive(Shrinkwrap, Debug, Default)]
#[shrinkwrap(mutable)]
pub(crate) struct DiagInterner {
    /// Store containing all diagnostics interned so far. Multiple stores can
    /// coexist in different interners, e.g., when parsing, and may be merged
    /// into one interner.
    pub(crate) store: Vec<Diagnostic>,
}

impl DiagInterner {
    /// Add a new diagnostic.
    pub(crate) fn intern(&mut self, d: impl Into<Diagnostic>) {
        self.store.push(d.into());
    }

    /// Merges two sets of diagnostics.
    pub(crate) fn merge(&mut self, mut other: Self) {
        self.store.append(&mut other.store)
    }

    /// Returns `true` if the interner is empty, otherwise `false`.
    pub(crate) fn is_empty(&self) -> bool {
        self.store.is_empty()
    }

    /// Drains all diagnostics of the interner and places them into a new interner.
    pub(crate) fn take(&mut self) -> Self {
        std::mem::take(self)
    }

    /// Emits all diagnostics in the interner. The HIR does not need to be passed if the
    /// diagnostics were generated only while parsing.
    pub(crate) fn emit<'i, W>(
        &self,
        info: &Info,
        hir: impl Into<Option<&'i HIR>>,
        f: &mut W,
    ) -> CodespanResult
    where
        W: Write + WriteColor,
    {
        // Emit header message
        f.set_color(ColorSpec::new().set_fg(Some(Color::Red)));
        writeln!(f, "[-- Found {} errors --],", self.len())?;
        tracing::trace!("{:?}", self);
        f.reset();

        let files = &info.files.store;
        let config = &Config::default();
        let ctx = &Context::new(info, hir.into());
        self.iter()
            .filter_map(|diag| diag.to_codespan(ctx))
            .try_for_each(|diag| term::emit(f, config, files, &diag))
    }
}

/// A diagnostic result.
pub(crate) type Result<T> = std::result::Result<T, Diagnostic>;

/// A compile-time or runtime diagnostic.
#[derive(Debug, From)]
pub enum Diagnostic {
    /// A compile-time note
    Note(Note),
    /// A compile-time warning
    Warning(Warning),
    /// A compile-time error
    Error(Error),
}

/// Compile-time info reported by the compiler.
#[derive(Debug)]
pub enum Note {}

/// Compile-time warnings reported by the compiler.
#[derive(Debug)]
pub enum Warning {}

/// Compile-time errors reported by the compiler.
#[derive(Debug)]
pub enum Error {
    /// Error when the importer fails to find a source file.
    FileNotFound,

    /// Error when the importing is given a file with an invalid extension.
    BadExtension,

    /// Error when the lexer encounters too many levels of indentation.
    TooMuchIndent {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when the lexer encounters indentation indivisible by TABSIZE.
    BadIndent {
        /// Location of the error.
        loc: Loc,
    },

    /// Error produced by [`lexical_core`].
    LexicalCore {
        /// Error code.
        err: lexical_core::Error,
        /// Location of the error.
        loc: Loc,
    },

    /// Error produced by [`time`].
    Time {
        /// Error code.
        err: time::ParseError,
        /// Location of the error.
        loc: Loc,
    },

    /// Error when the parser comes across an extra token.
    ExtraToken {
        /// Extra token found while parsing.
        found: Token,
        /// Location of the error.
        loc: Loc,
    },

    /// Error when the parser comes across an invalid token.
    InvalidToken {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when the parser comes across an unrecognized end-of-file.
    UnrecognizedEOF {
        /// Location of the error.
        loc: Loc,
        /// List of tokens expected by LALRPOP.
        expected: Vec<String>,
    },

    /// Error when the parser comes across an unrecognized token.
    /// In other words, a token emitted by the lexer which the parser
    /// did not expect.
    /// NB: This is technically an Internal-Compiler-Error.
    UnrecognizedToken {
        /// Unrecognized token found while parsing.
        found: Token,
        /// Location of the token.
        loc: Loc,
        /// List of tokens expected by LALRPOP.
        expected: Vec<String>,
    },

    /// Error when two types fail to unify.
    TypeMismatch {
        /// First type.
        lhs: Type,
        /// Second type.
        rhs: Type,
        /// Location of the error.
        loc: Loc,
    },

    /// Error when type information is needed at a code location, but is not supplied.
    TypeMustBeKnownAtThisPoint {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when a path does not reference anything.
    PathNotFound {
        /// Path which failed to resolve.
        path: Path,
        /// Location of the error.
        loc: Loc,
    },

    /// Error when a match is non-exhaustive.
    NonExhaustiveMatch {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when two items in the same namespace have the same name.
    NameClash {
        /// Name of the two items.
        name: Name,
    },

    /// Error when a struct contains two fields with the same name.
    FieldClash {
        /// Name of the two fields.
        name: Name,
    },

    /// Error when an enum contains two variants with the same name.
    VariantClash {
        /// Name of the two variants.
        name: Name,
    },

    /// Error when an enum variant is constructed with too many arguments.
    VariantWrongArity {
        /// Path of the enum variant.
        path: Path,
    },

    /// Error when a tuple is indexed with an out-of-bounds index.
    OutOfBoundsProject {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when a struct-field is accessed which does not exist.
    FieldNotFound {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when placing a type in value position.
    TypeInValuePosition {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when placing a value in type position.
    ValueInTypePosition {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when enwrapping a non-variant path.
    PathIsNotVariant {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when placing a refutable pattern where an irrefutable one is expected.
    RefutablePattern {
        /// Location of the error.
        loc: Loc,
    },

    /// Error when moving a used value, and the moved value does not implement Copy.
    UseOfMovedValue {
        /// Location of the parent value.
        loc0: Loc,
        /// Location of the second value.
        loc1: Loc,
        /// Type which does not implement Copy.
        t: Type,
    },

    /// Error when using the same value twice, and the used value does not implement Copy.
    DoubleUse {
        /// Location of the place expression declaration.
        loc0: Loc,
        /// Location of the first use.
        loc1: Loc,
        /// Location of the second use.
        loc2: Loc,
        /// Type which does not implement Copy.
        t: Type,
    },

    /// Error when an extern function contains a parameter which is a pattern.
    PatternInExternFun {
        /// Location of the extern function.
        loc: Loc,
    },

    /// Error when an expression expects a selector (e.g., e1 not in e2[123]), but the selector is
    /// not specified.
    ExpectedSelector {
        /// Location of the expression
        loc: Loc,
    },

    /// Error when multiple selectors are specified. For now this is a hard-error, but will relaxed
    /// in the future.
    MultipleSelectors {
        /// Location of the expression
        loc: Loc,
    },

    /// Error when using a non-selectable type (map, set) in a selector.
    ExpectedSelectableType {
        /// Location of the expression
        loc: Loc,
    },
}

/// Runtime errors reported by the compiler.
#[derive(Debug)]
pub enum Panic {
    /// Stack-trace to code location which caused a panic.
    Unwind {
        /// Location of the runtime error.
        loc: Loc,
        /// Path to each function in the call-stack.
        trace: Vec<Path>,
    },
}
