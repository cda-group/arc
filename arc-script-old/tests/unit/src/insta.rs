#![cfg_attr(not(test), allow(dead_code, unused_imports))]

use arc_script_compiler::prelude::diags::Buffer;
use arc_script_compiler::prelude::modes::{Input, Mode, Output};

use lazy_static::lazy_static;
use regex::Regex;

use std::env;

lazy_static! {
    static ref FILE_LOC_REGEX: Regex = Regex::new(r#"/.*:[0-9]+:[0-9]+"#).unwrap();
}

fn snapshot(output: Output, paths: &str) {
    env::set_var("INSTA_OUTPUT", "summary");
    env::set_var("INSTA_FORCE_PASS", "1");

    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_snapshot_suffix();
    let mut sink = Buffer::no_color();
    settings.bind(|| {
        insta::glob!(paths, |path| {
            println!("Testing {}", path.display());
            let mode = Mode {
                input: Input::File(Some(path.into())),
                output,
                no_prelude: true,
                ..Default::default()
            };
            arc_script_compiler::compile(mode, &mut sink).unwrap();
            let s = std::str::from_utf8(sink.as_slice()).unwrap();
            let s = FILE_LOC_REGEX.replace_all(s, regex::NoExpand("<FILE-NAME>"));
            insta::assert_snapshot!(s);
            sink.clear();
        })
    });
}

#[test]
fn test_ast() {
    snapshot(Output::AST, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::AST, "tests/expect_pass/*.arc");
    snapshot(Output::AST, "tests/expect_fail/*.arc");
    snapshot(Output::AST, "tests/expect_fail_todo/*.arc");
}

#[test]
fn test_hir() {
    snapshot(Output::HIR, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::HIR, "tests/expect_pass/*.arc");
    snapshot(Output::HIR, "tests/expect_fail/*.arc");
    snapshot(Output::HIR, "tests/expect_fail_todo/*.arc");
}

#[test]
fn test_rust() {
    snapshot(Output::Rust, "tests/expect_mlir_fail_todo/*.arc");
    snapshot(Output::Rust, "tests/expect_pass/*.arc");
    snapshot(Output::Rust, "tests/expect_fail/*.arc");
}

#[test]
fn test_mlir() {
    snapshot(Output::MLIR, "tests/expect_pass/*.arc");
    snapshot(Output::MLIR, "tests/expect_fail/*.arc");
}
