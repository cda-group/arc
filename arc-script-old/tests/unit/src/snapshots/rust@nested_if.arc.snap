---
source: tests/unit/src/insta.rs
expression: s
input_file: tests/unit/src/tests/expect_pass/nested_if.arc

---
#[allow(non_snake_case)]
#[allow(unused_must_use)]
#[allow(dead_code)]
#[allow(unused_variables)]
#[allow(unused_imports)]
#[allow(unused_braces)]
#[allow(irrefutable_let_patterns)]
#[allow(clippy::redundant_field_names)]
#[allow(clippy::unused_unit)]
#[allow(clippy::double_parens)]
pub mod arc_script_output {
    use super::*;
    use arc_script::codegen;
    use arc_script::codegen::*;
    pub fn test() -> i32 {
        let x_0: i32 = 65i32;
        let x_1: i32 = 66i32;
        let x_2: i32 = 67i32;
        let x_3: bool = true;
        let x_4: bool = false;
        let x_6: i32 = if val!(x_3) {
            val!(x_0)
        } else {
            let x_5: i32 = if val!(x_4) { val!(x_1) } else { val!(x_2) };
            val!(x_5)
        };
        val!(x_6)
    }
}
pub use arc_script_output::*;


